doc/HowToUseMeson.md

```diff
@@ -0,0 +1,234 @@
+         src="https://mesonbuild.com/assets/images/meson_logo.png" alt="meson source">
+    </a>
+</p>
+
+# Table of contents
+
+1. [Introduction](#introduction)
+2. [Project structure](#paragraph1)
+3. [Top level meson file](#paragraph2)
+    1. [Project settings](#subparagraph21)
+    2. [Useful system checks](#subparagraph22)
+4. [Meson + Python](#paragraph3)
+5. [Meson + Cython](#paragraph4)
+5. [How to compile apps, which are using mc_lib with meson](#paragraph5)
+
+---------
+## <a name="introduction"></a> Introduction 
+
+Some introduction. I'll add it later
+
+
+## <a name="paragraph1"></a> Project structure 
+
+All `meson.build` files connected between each other, it means,
+that variables deaclared in upper `meson.build` are visible
+from every subdirectorys `meson.build` files.
+
+The order of subdir() links is important for project. 
+
+Example:
+```
+project
+│   file001.pyx
+│   meson.build (A) 
+│   
+└───subdir1
+│   │   file011.pyx
+│   │   meson.build (B)
+│   │
+│   └───subdir1_1
+│       │   file111.pyx
+│       │   file112.pxd
+│       │   meson.build (C)
+│   
+└───subdir2
+    │   file021.pyx
+    │   meson.build (D)
+```
+
+`meson.build (A)`:
+
+```meson
+subdir('subdir1')
+subdir('subdir2')
+```
+Variables declared in `subdir1` are visible in `subdir2`, but not vice versa.
+
+
+## <a name="paragraph2"></a> Top level `meson.build`
+
+
+### <a name="subparagraph21"></a> Project settings 
+There should be declared:
+1. Project name,
+2. Programming languages(meson trying to find a suitable compiler for named languages ),
+3. Suitable meson version,
+4. App version,
+5. Licenses,
+6. Etc.
+
+Cython files compiles in order: `.pyx` &rarr; `.c` or `.cpp` &rarr; `.so`.
+So there must be declared at least 2 languages: `cpp` or `c` and `cython`.
+
+
+### <a name="subparagraph22"></a> Useful system checks 
+For apps which are using `cython` can be added manual check
+using `find_program` function:
+```meson
+cython = find_program('cython', required : true)
+if not cython.found()
+  error('MESON_BUILD_FAILED: Cython3 not found.')
+endif
+```
+
+Also, the `python` installation must be checked too. 
+Default `pyhton` dependencies saved into variable
+to insert them into every compile target
+```meson
+py_mod = import('python')
+py3 = py_mod.find_installation()
+py3_dep = py3.dependency()
+```
+As for `MC_lib`, the most part of it uses `numpy` API,
+so numpy declared as dependency, by its location:
+```meson
+incdir_numpy = run_command(py3,
+  ['-c', 'import os; os.chdir(".."); import numpy; print(numpy.get_include())'],
+  check : true
+).stdout().strip()
+
+inc_np = include_directories(incdir_numpy)
+```
+
+
+## <a name="paragraph3"></a> Meson + Python 
+There is no need to compile pure `python` files,
+so all of them can be installed as sources:
+```meson
+python_sources = [
+    '__init__.py',
+    'Pytester.py',
+] # declare list of .py files
+
+py3.install_sources(
+    python_sources,
+    pure: false,
+    subdir: install_dir
```

Comment:
This `install_dir` is incorrect. `py3.install_sources` already knows where to put things. The base dir to install is in the correct `site-packages` dir. So if you want to install into the root of your own package, then you use the package import name here (e.g. `mc_lib`). And for a subdir: `mc_lib/subdir1`

Reply:
Does this also apply to `py3.extension_module`?

Reply:
As I noted further down, you redefined `install_dir` so it's not technically incorrect, but it seems more robust to just use strings directly here. And yes, that applies to `py3.extension_module` too.